p2p/protocol/identify/peer_loop.go

```diff
@@ -0,0 +1,265 @@
+package identify
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/identify/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+)
+
+var errProtocolNotSupported = errors.New("protocol not supported")
+var isTesting = false
+
+type peerHandler struct {
+	ids *IDService
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	pid peer.ID
+
+	msgMu         sync.RWMutex
+	idMsgSnapshot *pb.Identify
+
+	pushCh     chan struct{}
+	deltaCh    chan struct{}
+	evalTestCh chan func() // for testing
+}
+
+func newPeerHandler(pid peer.ID, ids *IDService, initState *pb.Identify) *peerHandler {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ph := &peerHandler{
+		ids:    ids,
+		ctx:    ctx,
+		cancel: cancel,
+		pid:    pid,
+
+		idMsgSnapshot: initState,
+
+		pushCh:  make(chan struct{}, 1),
+		deltaCh: make(chan struct{}, 1),
+	}
+
+	if isTesting {
+		ph.evalTestCh = make(chan func())
+	}
```

Comment:
(ok, maybe I was too optimization eager. I won't object if you want to revert this.

Reply:
@Stebalien I'm good with this change. It's one channel alloc for every peer we are connected to.